Work In Progress - Toc v5.3.x rewrite

pmario opened the pull request:
This PR was created to create a NETLIFY Preview

**Edit**: Test build now works: https://deploy-preview-7961--tiddlywiki-previews.netlify.app/

### All TOC-macros Rewritten

The toc-macro now use **toc-include**, **toc-filter** and **toc-hide** fields as [discussed at Talk: ](https://talk.tiddlywiki.org/t/toc-macros-rewritten-a-lot-of-new-fuctionality-part-2/9071)

[Initial discussion at Talk (Part1)](https://talk.tiddlywiki.org/t/proposal-toc-macros-rewrite-using-if-functions-and-procedures-new-fuctionality/9046/6?u=pmario)

### CTRL-click Chevron Expands / Folds Tree Branches

- CTRL-clicking on a chevron should expand the whole tree
- CTRL-click again will close the whole tree again

**Advantage:** If everything is expanded, a "normal" click only fold the "current" branch

### New Node Fields

New tiddler fields allow the new macros to have fine-grained control over toc branches on a "per node" / "per branch" level. 

- `toc-filter`
   - Expects a **titlelist** or a **filter**
   - If a toc node contains a field named: `toc-filter`, that filter will **replace the standard filter** that normally creates the "child list" of that node. 
   - **nodes added** with toc-filter have a CSS **class="toc-item-filter"**
   - If you want to list nodes with the current tag you should add `[tag<currentTidder>!has[draft.of]]` if you do not want to list draft-tiddlers
   - **If there is no toc-filter**, the default filter is: `[all[shadows+tiddlers]tag<tag>!has[draft.of]$(sort)$]` where the last element is a text-substitution of the _sort_ parameter of the toc macro. 
<!--

-->
- `toc-include`
  - Expects a **titlelist** of a **filter**.
  - Tiddlers **will be added** to the existing accumulated list. 
  - node added with toc-include have a CSS **class="toc-item-include"**
<!--

-->
- `toc-hide`
  - If set to **yes**, the node will be removed from the child list
  - toc-hide **takes precedence** over all other fields. 
  - Expects a **titlelist** or a **filter**
  - If set to a filter-run  eg: `:and[prefix[todo]]` ... all accumulated tiddlers prefixed `todo` will be **removed**
  - Be aware: Without the `:and` filter-run-prefix, the filter `[prefix[todo]]` is equivalent to `[all[tiddlers]prefix[todo]]`
   - **parents** with a toc-hide field have a CSS **class="toc-item-hide"**
   - toc-hide nodes are internally evaluated in the same way as the _exclude_ parameter of the toc macro
<!--

-->
- `toc-itemTemplate`
   - Expects a template tiddler-title
   - If itemTemplate is set, it uses the content of the template tiddler to display a toc node
      - If itemTemplate is used captionField (see info below) is set to `text`

All **titlelists** of **filters** added to the new fields are internally evaluated using `[subfilter[tf.toc-*]]`

The field names used are defined with global variables. So for expert users, they can be overwritten like this. 

```
\procedure toc-hideField()    toc-hide
\procedure toc-includeField() toc-include
\procedure toc-filterField()  toc-filter

<div class="tc-table-of-contents">
<<toc tag:"Contents">>
</div>
```

------

### New Macro Parameters

- **captionField** (default: caption)
  - Expects a field-name. If set it uses this field instead of the caption-field
  - If the field is empty, or does not exists, it tries to use "caption" or falls back to the "tiddler-title"

- **tagField**
  - Uses any "tag-like" field instead of the "tags" field to create the "tag-like" tree
  - tagField takes precedence over parentField

- **parentField**
  - Instead of using tags, the toc tree is built using "parent" fields. 
  - There is no default, but usually it will be "parent"
  - This function uses the same filter behaviour as the tocP plugin

- **itemTemplate**
  - Expects a: template tiddler-title
  - by default toc macros use the `toc-defaultItemTemplate` procedure, which is specific to toc-macros
  - If itemTemplate is set, it uses the content of the template tiddler to display a toc node
     - If itemTemplate is used captionField is set to `text`
  - **Important:** If the `toc-itemTemplate` field is set in the current node it will take precedence over the macro parameter

The new fields should work with every toc-* macro.

-------

<details><summary>Test Styles are part of the JSON test-suite linked below ;)</summary>

```
.toc-item-filter> a,
.toc-item-filter> .tc-toc-caption {
  color: darkviolet;
}

.toc-item-filter> button > svg {
  fill: darkviolet;
}

.toc-item-include  > a,
.toc-item-include > .tc-toc-caption
{
  color: green;
}

.toc-item-include > button > svg {
  fill: green;
}

.toc-item-hide> a,
.toc-item-hide > .tc-toc-caption
{
  color: red;
}

.toc-item-hide> button > svg {
  fill: red;
}

```

</details>

-----

Requested (but not implemented yet) features from Part-1:

- Use a `$:/config/toc/`_currentTidder_ configuration to allow "per node" parameter change, whithout "touching" a user-created toc
  - This will allow plugin authors, or admins to modify toc-behaviour without the need to edit user-content.
- (If I forgot something -- let me know, below)

-----


### Test Data

The JSON attachment contains some `test-*` tiddlers that I use for testing atm. They are **chaotic**, but contain a bit of info and some links

* blue links are standard
* green links are `toc-include`d
* dark violet is `toc-filter`ed

JSON file can be found at: https://talk.tiddlywiki.org/t/toc-macros-rewritten-a-lot-of-new-fuctionality-part-2/9071?u=pmario

pmario committed:
replace <$qualify name="toc-state" title={{{ xxx filter }}} > with

\function makeTocState() [[$:/state/toc]addsuffix<__path__>addsuffix[/]addsuffix<currentTiddler>]

<$qualify name="toc-state" title=<<makeTocState>> >

pmario committed:
toc change sorted example to  use "!sort[title]"

pmario committed:
toc rewrite wip - works - before improving emptymessages

pmario committed:
toc rewrite wip - works - replaced emptyMessages with conditional if

pmario committed:
toc rewrite -- clean up the code and add comme important comments

pmario committed:
toc - replace reveal with conditional if

pmario committed:
toc - bring docs in line with the code

pmario committed:
add internal variable prefixes to __exclude and __path to make the code less error prone

pmario committed:
update docs for TOC and examples

pmario committed:
toc-exclude works, but should be improved

pmario commented:
To look at Files Changed at GitHub. Use the diff  "split" setting to see what's going on

![image](https://github.com/Jermolene/TiddlyWiki5/assets/374655/cc23bb0e-b99d-45b0-8874-e86a5a9b8f2f)

pmario commented:
The existing "Table-of-Contents Macros" tiddler got an overhaul, to make it more consistent and easier to understand

pmario committed:
use 2 underscores instead of 1 for better readability

pmario committed:
add toc-include, toc-exclude to all tocs

pmario committed:
add new classes toc-item-included and toc-item-excluded, simplify code using more functions

pmario committed:
show chevron if toc-included is the only "child" of a node. some code restructuring for better readability

pmario committed:
toc-include replaces the default filter

pmario committed:
change possible fields to: toc-hide, toc-override and toc-include

pmario committed:
add class handling for toc-override

pmario committed:
rename toc-override -> toc-filter

pmario committed:
fix typo in class function

pmario committed:
add global field names

pmario committed:
add global field names

pmario committed:
Merge branch 'toc-v5.3.x-rewrite' of github.com:pmario/TiddlyWiki5 into toc-v5.3.x-rewrite

pmario committed:
fix Second

pmario committed:
remove test-files

pmario committed:
toc examples use named params now

pmario committed:
add viewField instead of captions, use viewFieldTemplate if defined to show item-link text. It always uses field:text

pmario committed:
fix let syntax

pmario committed:
remove itemClassFilter from macro-calls, but leaf them alone in the procedure definitions. So parameter order is kept intakt

pmario committed:
docs - move exclude parameter into Parameters section

pmario committed:
ren: viewField to captionField,viewFieldTemplate to captionFieldTemplate

pmario committed:
allow tagField and parentField instead of tags

pmario committed:
rename toc-captionField to toc-defaultItemTemplate and lists call toc-itemTemplate which is a new proc

pmario committed:
add all new fields to all toc-macros. rename some internal variables

pmario committed:
doc - add accidentally removed fields

pmario committed:
move captionField in front of tagField

pmario committed:
improve inline docs

pmario committed:
docs - adjust TOC Example tiddler

pmario committed:
docs -- add tooltip field info

pmario committed:
WIP recursive open - close with tags and path works with CTRL click and buttons

pmario committed:
CTRL click on chevron expands / folds whole branch

pmario committed:
if currentTiddler includes itself as a child do not show a more-chevron

pmario committed:
CTRL-click should not create edundant state tiddlers if tag tiddlers include themself

pmario committed:
allow toc-filter fields to created open/close chevrons

pmario commented:
@Jermolene -- Do you think these new macros will have a chance to be merged into the core?

pmario commented:
@Jermolene -- The OP at GH now contains the "docs" info copied over from the talk thread part 2

pmario committed:
updated inline docs

Jermolene commented on core/wiki/macros/toc.tid:
```diff
@@ -1,218 +1,422 @@
 tags: $:/tags/Macro
 
+<!--
+Naming rules / pattern
```
Hi @pmario my main concern at the moment is with the comments that make up the documentation. They are written in a style that is brief and abbreviated (for example, why use a slash in the line above?) and therefore unnecessarily hard to understand.

I am not a fan of inventing new symbols like `->`. The important thing for a project like TiddlyWiki is consistency, and so it is not helpful if one file full of macros adopts different standards for documentation.

pmario replied:
will change it

Jermolene commented on core/wiki/macros/toc.tid:
```diff
@@ -1,218 +1,422 @@
 
-\define toc-open-icon() $:/core/images/down-arrow
-\define toc-closed-icon() $:/core/images/right-arrow
+The toc macro code gets more and more complex from top to bottom.
```
"callse" typo

Jermolene commented on core/wiki/macros/toc.tid:
```diff
@@ -1,218 +1,422 @@
+The toc macro code gets more and more complex from top to bottom.
+toc-tabbed-internal-nav calls toc-tabbed-external-nav, which callse toc-selective-expandable, which calls toc-expandable
+
+The very first toc macro is _not_ "reused". So it can be used as a "simple" reverence implementation,
```
The use of double quotes is unusual and a bit clumsy. What are they trying to accomplish?

pmario replied:
They give me a visual hint about the imortant words in the text, since comments cannot be formatted.

Jermolene commented on core/wiki/macros/toc.tid:
```diff
@@ -1,218 +1,422 @@
+toc-tabbed-internal-nav calls toc-tabbed-external-nav, which callse toc-selective-expandable, which calls toc-expandable
+
+The very first toc macro is _not_ "reused". So it can be used as a "simple" reverence implementation,
+because it is realatively easy to debug.
```
"realatively" typo

Jermolene commented on core/wiki/macros/toc.tid:
```diff
@@ -1,218 +1,422 @@
+</$transclude>
+\end
+
+<!-- v5.2.4 add "Custom Icons" -->
```
What does this comment mean? Is it just marking something that was added in v5.2.4? Why is it needed?

pmario replied:
I did add this info to see, that I cannot change those names for backwards compatibility reasons. Otherwise I may have changed them by accident

I'll change it to: `<!-- Allow custom icons (TW v5.2.4). Keep names for backwards compatibility -->`

Jermolene commented on core/wiki/macros/toc.tid:
```diff
@@ -1,218 +1,422 @@
+<!-- join the CSS definitions from above -->
+\function tf.toc-itemClass()       [<tf.toc-itemClassFilter>] [<tf.toc-includeClass>] [<tf.toc-filterClass>] [<tf.toc-hideClass>] +[join[ ]]
+
+<!-- Special case, if toc-hide field is "yes", we need a currentTiddler *titlelist* -- otherwise return the content as text -->
```
I don't understand what this sentence is trying to communicate but the use of commas here is wrong. Better to say:

```
Special case: if the toc-hide field is "yes", we need a currentTiddler *titlelist* -- otherwise return the content as text
```

pmario replied:
Will change it.

More info: toc-hide is special. It can be set to "yes" for ease of use, or it can be a filter. Rudimentary docs for the "toc-hide" field is at: https://talk.tiddlywiki.org/t/toc-macros-rewritten-a-lot-of-new-fuctionality-part-2/9071

Jermolene commented on core/wiki/macros/toc.tid:
```diff
@@ -1,218 +1,422 @@
-      </$reveal>
-    </li>
-  </$set>
+<!-- ==========================================================
```
This is another non-standard new type of marker. We should be consistent.

pmario replied:
Yes. I think we should introduce something like this when we update core macros. It makes it much easier to visually separate the main macros from helper functions.

Jermolene commented on core/wiki/macros/toc.tid:
```diff
@@ -1,218 +1,422 @@
+		tagField=<<tagField>>
+		parentField=<<parentField>>
+		itemTemplate=<<itemTemplate>>
+	/>
```
This file seems like it is now unmanageably large.

pmario replied:
I did reply to this one, in the main thread

Jermolene reviewed (commented):
Thanks @pmario this is a huge and welcome piece of work. I have some concerns about the size of the file, and I think the documentation comments need a little more work. I wonder if you could collaborate with another contributor on this?

pmario commented:
>@jermolene wrote
>
>This file seems like it is now unmanageably large.

As you can see in the <$macrocall $name="toc-tabbed-... directly above your comment. It uses 12 lines now instead of 1. IMO those 12 lines make the code much more readable and manageable. 

IMO the same is true for all other $marcrocalls and all $list-widget filters. They mainly use 4 lines now instead of 1. 
Changing that back would make the file 42 lines "shorter" but much harder to read for other mainainers than me. 

The file now contains 50+ lines of inline comment, which IMO also makes maintenence easier, since they point out, where to pay attention. 

The original toc-macros tiddler had about 220 lines of code. The new macors have about 300 real lines of code, but they also have a lot more functionality. 

- The toc-macros can now handle tag-based, field-based and parent-based tocs using some new macro parameters
- It can handle: per item-based "hide", "include" and "filter" fields for every one of the above new macro "types"
- It can handle: open / close a "full branch" with CTRL-click on the "more chevron"
- It also allows to create open / close "full branch" buttons added to a custom item-template
- It can handle: global item templates to show other info than caption or the tiddler title
- It can handle: custom templates on a "per item" basis 
- With the right templates toc-macros can now handle a tree that was created by the streams-plugin from Saq
- With the right settings and templates, they can handle trees created by my tocP-plugin, which can handle "parent based" tocs and Saq's streams

All that new functionality could be achieved by about a 100 lines of new code and a lot of code-reuse with some new settings.
I think it would have not been possible to do this with the "old" syntax. Especially \functions work great now.

Most of the ideas discussed at Talk: https://talk.tiddlywiki.org/t/proposal-toc-macros-rewrite-using-if-functions-and-procedures-new-fuctionality/9046/34 could be implemented with 100 lines of code. 

I think maintaining the 300 lines of code is much easier than having 4 files with 200 lines each, where 90% of the code is the same. Everytime the core changes 1 of them all 4 of them would need to be changed. 

Just my thoughts

pmario committed:
change inline docs. Move \function tf.toc-stateTitle() up in the code near it's first use

pmario commented:
@Jermolene .. The OP of this thread: https://github.com/Jermolene/TiddlyWiki5/pull/7961#issue-2116618606 also contains the preliminary docs of the new macro params and item fields. 

The current docs and the examples can be [tested with the latest Vercel preview](https://tiddlywiki5-git-fork-pmario-toc-v53x-rewrite-jermolene.vercel.app/#Table-of-Contents%20Macros:%5B%5BTable-of-Contents%20Macros%5D%5D%20%5B%5BTable-of-Contents%20Macros%20(Examples)%5D%5D)

The new paramters and item-fields have not been documented with TW yet.

pmario commented:
The test-date I use atm is this one. [toc-macros-new-test-tiddlers.zip](https://github.com/Jermolene/TiddlyWiki5/files/14229900/toc-macros-new-test-tiddlers.zip)

pmario committed:
toc-macros - somehow a "dot" slipped in with a commit :/

pmario commented:
>Thanks @pmario this is a huge and welcome piece of work. 

Thanks

>I have some concerns about the size of the file,

As I wrote. About 100 lines come from more readable macro-calls and inline docs. About 100 lines are new functionality. 

>and I think the documentation comments need a little more work. 

That's always true. ATM the inline docs makes sense for me, but may cause questions for others.

> I wonder if you could collaborate with another contributor on this?

Collaborate on the inline docs or the macros?

pmario committed:
remove draft tiddler

Jermolene commented:
Hi @pmario 

> Collaborate on the inline docs or the macros?

I meant the docs. We are exceptionally lucky to have so many willing contributors. The linguistic diversity of the community is a great strength, too, and helps us avoid the cardinal mistake of unknowingly creating English language-centric software.

There is however a higher barrier for documentation from contributors for whom English is not their first language. It generally falls to me to help contributors get the docs into shape, which is of course our usual bottleneck. Meanwhile, there are no shortage of members of the community who have the necessary English skills and so might be in a position to help.

pmario committed:
toc - add tc- prefix to class definitions, to make them more obvious

pmario committed:
toc - do not show "more" chevron, it all child nodes are hidden

pmario committed:
toc - add new preliminary tests

pmario committed:
rename 1 variable change docs a bit

pmario committed:
Some new examples - to be explained in detail

pmario commented:
>Roma a PM at Talk
>
>I noticed that the menubar’s Contents dropdown disappears when I interact with any of the expand/collapse chevron buttons. The click clearly succeeds (when I re-invoke the dropdown, the submenus show/hide in modified way, as expected), but the current version fails to leave the drop-down open for further interaction after the chevron is clicked.
>
>No need to reply, unless you don’t see what I see and there’s more detail I can offer to help troubleshoot.

pmario commented:
Investigate this one:

>Mentioned at Talk: https://talk.tiddlywiki.org/t/optimal-toc-use-for-multiple-topics/9804/10?u=pmario
>
>I have noticed that the captionField does not seem to use the caption field in the absense of the new field set, I am unsure if this is just because I copy/pasted the $:/core/macros/toc tiddler to a blank tw or not, but I felt I should mention it here.

pmario commented:
>https://talk.tiddlywiki.org/t/optimal-toc-use-for-multiple-topics/9804/14?u=pmario
>
>... would it be difficult to adda tooltipField for example tooltipField=description would display as the tooltip the content of the description field on hover?, if using the caption they could provide the title as the tooltip field, or tooltiop etc…
>
> - The link widget (now) and Button widget (for a while) have the tooltip parameter, this can be a way to introduce other information, especlialy if its transcluded.

pmario committed:
add a tooltip and make caption the default fallback

pmario committed:
update test tiddlers

pmario committed:
add cascade to check if macro prams and fields are valid. eg Caption has a value

pmario committed:
fix chevron spacing for custom itemTemplates

pmario committed:
add tf.toc-getTooltip function

pmario committed:
Merge branch 'master' into toc-v5.3.x-rewrite

pmario committed:
Merge branch 'master' into toc-v5.3.x-rewrite

pmario committed:
add toc-newChildButton which uses tv-toc-show-new-child variable for visibility

pmario committed:
new test tiddlers

pmario committed:
fix toc new child - titlecase problem

pmario committed:
fix problem tagMode new child also opens the tiddler

pmario committed:
fix TAG-LIKE functionality

pmario committed:
update all test-tiddlers with wikitext-example copy to clipboard function

pmario committed:
change test-tagLike-toc to wikitext-example

pmario committed:
allow tv-toc-show-new-child to be used with tagLikeMode

pmario committed:
Merge branch 'master' into toc-v5.3.x-rewrite

pmario committed:
make template a link

pmario committed:
remove button tc-tiny-gap from tc-unlinked-selective-expandable-body

pmario committed:
remove redundant tagLike procedures

pmario committed:
add 2 testcase examples for toc-simple

pmario committed:
use testcases for toc documentation

pmario committed:
add CSS classes to new-child buttons for easier styling

pmario committed:
Merge branch 'master' into toc-v5.3.x-rewrite

pmario committed:
merge master

pmario committed:
update examples

pmario committed:
Merge branch 'master' into toc-v5.3.x-rewrite

pmario committed:
add new-child core image

pmario committed:
update some ecampel test-cases

pmario committed:
wip toc testcase make new child button work

pmario committed:
add test-new-tiddler-button - GH-issue

pmario committed:
change to [prefix[$:/StoryList]], [prefix[$:/HistoryList]]

pmario committed:
Make testcase tab-view a story-river

pmario committed:
fix typo textcase -> testcase

pmario committed:
remove $:/StoryList and $:/HistoryList from Export filter

pmario committed:
rename story-river template

pmario committed:
merge master

pmario committed:
remove obsolete code

pmario committed:
update docs rename SimpleToc to /param/<parameterName>

pmario committed:
fix missingText in toc-tabbed-*-nav

pmario committed:
remove redundant comment

pmario committed:
use currentTiddler as tag default parameter if it is missing

pmario committed:
Merge branch 'master' into toc-v5.3.x-rewrite

pmario committed:
update a test

pmario committed:
merge master

pmario committed:
Merge branch 'master' into toc-v5.3.x-rewrite

pmario committed:
rename toc-item to toc-chevron

pmario committed:
rename toc-defaultItemTemplate to toc-defaultFieldTemplate

pmario committed:
move the link-widget into the _linkTemplate

pmario committed:
set _hasItemTemplate field to title instead of text

pmario committed:
_linkTemplate also handles toc-link[no]

pmario committed:
_linkTemplate also handles toc-link[no]

pmario committed:
introduce tf.toc-stateIsClosed() and tf.toc-stateIsOpen() functions to test toc trees fully expanded

pmario committed:
fix tests

pmario committed:
introduce tv-toc-expand() variable for fully expanded toc testing

pmario committed:
add tests for current examples

pmario committed:
Merge branch 'TiddlyWiki:master' into toc-v5.3.x-rewrite

pmario committed:
merge master

pmario committed:
Add the level parameter

pmario committed:
Add some basic manual simple toc tests

pmario committed:
Add some more toc tests

pmario committed:
Update the TableOfContentsMacro docs

pmario committed:
Add missing toc specs

pmario committed:
Add example matrix, so users can easily play with examples

pmario committed:
Update dev docs

pmario committed:
toc - adjust comments

pmario committed:
Update itemClassFilter docs

pmario committed:
Replace toc-linked-expandable-body and toc-unlinked-expandable-body
with single toc-expandable-body. Move the toc-link check inside

pmario committed:
The 2x2 matrix (has-children/no-children x linked/unlinked)
is now in a single toc-selective-expandable-body procedure,
with all button classes and tabindex differences preserved exactly.
The debug $log comments are kept with the <!-- Keep $log --> marker.

Key specs to verify:

toc-simple-selectiv-expandable,
toc-simple-selectiv-expandable-expanded,
toc-simple-selectiv-expandable-sort-title
toc-simple-selectiv-expandable-custom-icons
toc-advanced-selectiv-expandable-toc-link-no
toc-simple-tabbed-internal
toc-simple-tabbed-external.

pmario committed:
Improve selected styles

pmario committed:
Fix hover shift other nodes problem

pmario committed:
Update toc macors docs

pmario committed:
Update base vanilla CSS with rewritten TOC styles

pmario committed:
disable fix-toc-styles stylesheet - is merged to vanilla now

pmario committed:
rename toc examples part 1

pmario committed:
rename toc examples part 2

pmario committed:
move toc examples into their own directory

pmario committed:
Fix narrative and move to dedicated subdirectories

pmario committed:
more advanced toc examples

pmario committed:
Implement toc-branchTemplate that is used for all children of that branch

pmario committed:
toc-branchTemplate full test coverage

pmario committed:
fix typos and remove redundant test tiddlers

pmario committed:
Link Example Matrix and tag it

pmario committed:
remove redundant test tiddlers

pmario committed:
remove redundant test tiddlers

pmario committed:
remove redundant test tiddlers

pmario committed:
revert changes to Second

pmario committed:
add toc-branchTemplate to the docs

pmario committed:
prepare for testing

pmario committed:
prepare for testing

pmario committed:
add mcp server to toc rewrite worktree

pmario committed:
Merge branch 'master' into toc-v5.3.x-rewrite

pmario committed:
Split tw5.com static build to cut peak memory

Netlify previews OOM-kill the tw5.com build: one node process rendering
the site plus all static pages peaks at ~3.3GB.

- Add static-base, static-alltiddlers and static-tiddlers build targets;
  keep the static target unchanged for compatibility
- Prefix the two heavy targets with the index target's --setfield
  externalisation commands so they work standalone
- Run them as separate node processes in build-site.sh; worst
  per-process peak drops to ~1.9GB, output stays byte-identical

pmario commented:
@Jermolene ... I think I did find the problem, which causes test-build **fails** because of static rendering cause memory problems.

See: https://github.com/TiddlyWiki/TiddlyWiki5/pull/7961/commits/513ffea96c4617cb44bddc7b62ba46d457b1810e

I did create these changes and now the build process works again.

NETLIFY: Test build: https://deploy-preview-7961--tiddlywiki-previews.netlify.app/

pmario committed:
toc: fix nested sort, caption class and compat

- toc-body: re-pass sort to recursive calls; omitted procedure
  parameters reset to defaults, so nested levels rendered unsorted
- Restrict the legacy toc-caption override to tagMode as documented
- Give the default caption span its tc-toc-caption class
- Update Examples/TOC spec expectations for the class change; capture
  the two missing ExpectedResults

pmario committed:
jasmine: import fields for wiki-based tests

Testcases keeping their ExpectedResult in a separate compound never ran
a comparison and passed trivially.

- Import payload from the import and import-compound fields, like
  $:/core/ui/TestCaseTemplate; own subtiddlers still win
- Import global macros before rendering Output; unlike the browser
  testcase widget the runner has no page scope, so <<toc>> rendered empty

pmario committed:
toc docs: add examples, fix spec wiring

- New examples with specs: param/sort, param/template,
  advanced/custom-level-indicator
- Point 7 import-compound references at the actual spec titles;
  the dangling ones silently skipped comparison
- Capture missing expectations for tabbed-external, tabbed-internal,
  toc-hide-methods
- Parameters doc: correct the itemTemplate entry
